packages/pyright-internal/src/analyzer/program.ts

```diff
@@ -725,6 +723,28 @@ export class Program {
         }
     }
 
+    // This will allow the callback to execute a type evaluator with an associated
+    // cancellation token and provide a read-only program.
+    run<T>(callback: (p: ProgramView) => T, token: CancellationToken): T {
+        const evaluator = this._evaluator ?? this._createNewEvaluator();
+        return evaluator.runWithCancellationToken(token, () => callback(this));
+    }
+
+    createSourceMapper(
```

Comment:
we could change name to `getSourceMapper` rather than `create` since that seems implementation detail.

I looked to see whether I can actually take this one out of program, but since we want to load some shadow files in program, it has to be in program. but it is not actually mutating program in anyway that can impact `semantic result`, so it is part of `programView`